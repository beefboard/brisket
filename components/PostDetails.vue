<template>
  <div class="container">
    <div class="top">
      <div class="details">
        <nuxt-link :to="`/profiles/${post.author}`" class="author">{{ post.author }}</nuxt-link>
        <div class="date">{{ new Date(post.date) | moment("dddd, MMMM Do YYYY") }}</div>
      </div>
      <div class="images" v-if="images.length > 0">
        <gallery :images="images" :index="index" @close="index = null"></gallery>
        <div
          class="image"
          v-for="(image, imageIndex) in images"
          :key="imageIndex"
          @click="index = imageIndex"
          :style="{ backgroundImage: 'url(' + image + ')', width: '300px', height: '200px' }"
        ></div>
      </div>
    </div>
    <div class="title">{{ post.title }}</div>
    <div class="main">
      <div class="content">
        <div
          v-for="(paragraph, index) of post.content.split('\n')"
          :key="index">
          {{ paragraph }}
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import config from '~/nuxt.config'

export default {
  name: 'post-details',
  props: {
    post: {
      type: Object,
      required: true
    }
  },
  data() {
    const images = []

    for (let i = 0; i < this.post.numImages; i++) {
      images.push(
        `${config.axios.baseURL}/v1/posts/${this.post.id}/images/${i}`
      )
    }

    return {
      images: images,
      index: null
    }
  }
}
</script>

<style scoped>
.container {
  flex: 1;
  margin: 1rem;
}

.top {
  display: flex;
  flex-direction: row;
}

.main {
  display: flex;
  flex-direction: row;
  justify-content: center;
}

.content {
  max-width: 70rem;
  margin-top: 2rem;
  font-size: 1.2rem;
}

.content > div {
  margin: 1rem;
}

@media screen and (max-width: 600px) {
  .top {
    flex-direction: column;
  }
  .image {
    width: 7rem !important;
    height: 5rem !important;
  }
}

.title {
  font-size: 2.4rem;
  font-weight: bold;
  color: rgb(59, 59, 59);
  text-align: center;
}

.details {
  display: flex;
  flex-direction: column;
  flex: 1;
  font-size: 0.8rem;
}

.author {
  text-transform: capitalize;
}

.images {
  flex: 1;
  display: flex;
  justify-content: flex-end;
  flex-wrap: wrap;
}

.image {
  cursor: zoom-in;
  width: 15rem !important;
  height: 10rem !important;
  float: left;
  background-size: cover;
  background-repeat: no-repeat;
  background-position: center center;
  border: 1px solid #ebebeb;
  margin: 1rem;
  box-shadow: 0 3px 7px rgba(0, 0, 0, 0.25), 0 1px 2px rgba(0, 0, 0, 0.22);
}
</style>

