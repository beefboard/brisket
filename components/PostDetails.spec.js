import VueTestUtils from '@vue/test-utils'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import PostDetails from './PostDetails'
import VueMoment from 'vue-moment'
import config from '~/nuxt.config'
import flushPromises from 'flush-promises'

VueTestUtils.config.stubs['nuxt-link'] = '<a><slot /></a>'
VueTestUtils.config.stubs['nuxt'] = '<div><slot /></div>'
VueTestUtils.config.stubs['fa'] = '<div><slot /></div>'
VueTestUtils.config.stubs['gallery'] = '<div><slot /></div>'

describe('PostDetails', () => {
  let mockAuth

  function renderLayout(post) {
    const localVue = createLocalVue()
    localVue.use(VueMoment)

    if (!post) {
      post = {
        title: 'test',
        id: 'asdfasdf',
        content: 'sadfasdfsadf sdsdsdg',
        author: 'test',
        date: new Date(2017, 11, 28).toISOString(),
        approved: false,
        pinned: false,
        votes: { grade: 1 }
      }
    }
    return shallowMount(PostDetails, {
      localVue,
      propsData: {
        post: post
      },
      mocks: {
        $store: {
          dispatch: jest.fn(),
          state: {
            auth: mockAuth,
            baseURL: 'https://test.com'
          }
        },
        $router: {
          push: jest.fn(),
          beforeEach: jest.fn()
        }
      }
    })
  }

  beforeEach(() => {
    mockAuth = null
  })
  test('mounts properly', () => {
    const wrapper = renderLayout()
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  test('renders properly', () => {
    const wrapper = renderLayout()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should link author to author profile', () => {
    const wrapper = renderLayout({
      title: 'test',
      id: 'asdfasdf',
      content: 'sadfasdfsadf sdsdsdg',
      author: 'slfkgjsdklfg',
      date: new Date(2017, 11, 28).toISOString(),
      approved: false,
      pinned: false,
      votes: { grade: 1 }
    })

    expect(wrapper.find('.author').attributes('to')).toBe(
      '/profiles/slfkgjsdklfg'
    )
  })

  it('should render post date to page', () => {
    const wrapper = renderLayout({
      title: 'test',
      id: 'asdfasdf',
      content: 'sadfasdfsadf sdsdsdg',
      author: 'slfkgjsdklfg',
      date: new Date(2017, 11, 28).toISOString(),
      approved: false,
      pinned: false,
      votes: { grade: 1 }
    })

    expect(wrapper.find('.date').text()).toBe('Thursday, December 28th 2017')
  })

  it('should render title into page', () => {
    const wrapper = renderLayout({
      title: 'test',
      id: 'asdfasdf',
      content: 'sadfasdfsadf sdsdsdg',
      author: 'slfkgjsdklfg',
      date: new Date(2017, 11, 28).toISOString(),
      approved: false,
      pinned: false,
      votes: { grade: 1 }
    })

    expect(wrapper.find('.date').text()).toBe('Thursday, December 28th 2017')
  })

  it('should render title into page', () => {
    const wrapper = renderLayout({
      title: 'test',
      id: 'asdfasdf',
      content: 'sadfasdfsadf sdsdsdg',
      author: 'slfkgjsdklfg',
      date: new Date(2017, 11, 28).toISOString(),
      approved: false,
      pinned: false,
      numImages: 2,
      votes: { grade: 1 }
    })

    expect(wrapper.find('.date').text()).toBe('Thursday, December 28th 2017')
  })
})
