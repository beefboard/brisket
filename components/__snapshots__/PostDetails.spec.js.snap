// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostDetails renders properly 1`] = `
<div class="container">
  <div class="top">
    <div class="details"><a to="/profiles/test" class="author">test</a>
      <div class="date">Thursday, December 28th 2017</div>
      <div class="info">Awaiting approval</div>
    </div>
    <div class="images">
      <div images="https://test.com/v1/posts/asdfasdf/images/0,https://test.com/v1/posts/asdfasdf/images/1,https://test.com/v1/posts/asdfasdf/images/2"></div>
      <div class="image" style="background-image: url(https://test.com/v1/posts/asdfasdf/images/0); width: 300px; height: 200px;"></div>
      <div class="image" style="background-image: url(https://test.com/v1/posts/asdfasdf/images/1); width: 300px; height: 200px;"></div>
      <div class="image" style="background-image: url(https://test.com/v1/posts/asdfasdf/images/2); width: 300px; height: 200px;"></div>
    </div>
  </div>
  <div class="title">test</div>
  <div class="main">
    <div class="content">
      <div>sadfasdfsadf sdsdsdg</div>
    </div>
  </div>
</div>
`;
